### PR TITLE
OAK-10912 : make fullGc include/exclude paths configurable via OSGI config

### PR DESCRIPTION
**Pull request by rishabhdaim**

### CODE REVIEW

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/rdb/RDBDocumentNodeStoreBuilder.java [hunk 1]
```diff
@@ -128,6 +132,28 @@ public RDBDocumentNodeStoreBuilder setFullGCEnabled(boolean b) {
...
     }
 
+    @Override
+    public Set<String> getFullGCIncludePaths() {
+        return of();
+    }
+
+    @Override
+    public RDBDocumentNodeStoreBuilder setFullGCIncludePaths(@NotNull String[] includePaths) {
+        // fullGC is non supported for RDB
```
Comment (shodaaan) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/rdb/RDBDocumentNodeStoreBuilder.java]: minor typo: fullGC is not supported

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/rdb/RDBDocumentNodeStoreBuilder.java [hunk 2]
```diff
@@ -128,6 +132,28 @@ public RDBDocumentNodeStoreBuilder setFullGCEnabled(boolean b) {
...
+    }
+
+    @Override
+    public Set<String> getFullGCExcludePaths() {
+        return of();
+    }
+
+    @Override
+    public RDBDocumentNodeStoreBuilder setFullGCExcludePaths(@NotNull String[] excludePaths) {
+        // fullGC is non supported for RDB
```
Comment (shodaaan) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/rdb/RDBDocumentNodeStoreBuilder.java]: minor typo: fullGC is not supported

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreService.java [hunk 3]
```diff
@@ -499,6 +499,8 @@ private void configureBuilder(DocumentNodeStoreBuilder<?> builder) {
                 setDocStoreEmbeddedVerificationFeature(docStoreEmbeddedVerificationFeature).
                 setThrottlingEnabled(config.throttlingEnabled()).
                 setFullGCEnabled(config.fullGCEnabled()).
+                setFullGCIncludePaths(config.fullGCIncludes()).
```
Comment (shodaaan) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreService.java]: Please consider adding a check and WARN message if a path is part of both include and exclude lists. 
I am thinking exclude would take priority in this case and the path would be excluded, but it is something we can discuss if you think it is relevant.

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreServiceConfigurationTest.java [hunk 4]
```diff
@@ -115,6 +118,22 @@ public void fullGCEnabled() throws Exception {
         assertEquals(fullGCDocStore, config.fullGCEnabled());
     }
 
+    @Test
```
Comment (shodaaan) [inline oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreServiceConfigurationTest.java:121]: Please consider also testing the logic when a path element appears in both the includes and excludes list.
Depending on what we decide the logic should be in this case (I propose that exclude takes priority), the test should reflect that.

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/mongo/MongoDocumentNodeStoreBuilderTest.java [hunk 5]
```diff
@@ -60,6 +62,22 @@ public void fullGCFeatureToggleDisabled() {
         assertNull(builder.getDocStoreFullGCFeature());
     }
 
+    @Test
```
Comment (shodaaan) [inline oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/mongo/MongoDocumentNodeStoreBuilderTest.java:65]: Please consider unit testing for invalid paths - maybe use a RegEx or a Utils method that validates the path string.

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/rdb/RDBDocumentNodeStoreBuilderTest.java [hunk 6]
```diff
@@ -67,6 +68,20 @@ public void fullGCDisabled() {
         assertFalse(builder.isFullGCEnabled());
     }
 
+    @Test
```
Comment (shodaaan) [inline oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/rdb/RDBDocumentNodeStoreBuilderTest.java:71]: Same here, please consider unit testing for invalid paths.

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreService.java (hunk 3 above)
Comment (rishabhdaim) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreService.java:502]: That's a good catch. I have tested with a path in both the `include` & `exclude` list and it turns out that in that case, we won't be fetching any document from db.
cc @stefan-egli

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreServiceConfigurationTest.java (hunk 4 above)
Comment (rishabhdaim) [reply oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreServiceConfigurationTest.java:121]: I believe both lists should be different, with even one common path, nothing gets fetched from DB.
cc @stefan-egli

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreService.java (hunk 3 above)
Comment (rishabhdaim) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreService.java]: I have tested on local and in case a path is present in both `include` & `exclude`, it won't be selected.

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreServiceConfigurationTest.java (hunk 4 above)
Comment (rishabhdaim) [reply oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreServiceConfigurationTest.java:121]: > I believe both lists should be different, with even one common path, nothing gets fetched from DB. cc @stefan-egli

my above is wrong, only common path is excluded not all.

Comment (rishabhdaim) [reply oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreServiceConfigurationTest.java:121]: Check for invalid paths is added here: https://github.com/apache/jackrabbit-oak/pull/1551/commits/772c8559d4fdfa7e9fd8d251272f18606325ca04

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/rdb/RDBDocumentNodeStoreBuilderTest.java (hunk 6 above)
Comment (rishabhdaim) [reply oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/rdb/RDBDocumentNodeStoreBuilderTest.java:71]: added in https://github.com/apache/jackrabbit-oak/pull/1551/commits/167b46c5eaa518d39bad952853ac55bed2e0b7a5

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/mongo/MongoDocumentNodeStoreBuilderTest.java (hunk 5 above)
Comment (rishabhdaim) [reply oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/mongo/MongoDocumentNodeStoreBuilderTest.java:65]: added in https://github.com/apache/jackrabbit-oak/pull/1551/commits/167b46c5eaa518d39bad952853ac55bed2e0b7a5

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/Configuration.java [hunk 7]
```diff
@@ -260,6 +260,27 @@
                     "are separated with '::'. Example: -Doak.documentstore.persistentCacheIncludes=/content::/var")
     String[] persistentCacheIncludes() default {"/"};
 
+    @AttributeDefinition(
+            name = "Full GC Include Paths",
+            description = "Paths which should be included in full garbage collection." +
```
Comment (stefan-egli) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/Configuration.java]: formatting is broken as some spaces are missing after dot

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/Configuration.java [hunk 8]
```diff
@@ -260,6 +260,27 @@
...
+                    "Empty value means all paths are included. " +
```
Comment (stefan-egli) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/Configuration.java]: I find this potentially confusing.

Currently when no include path is specified, it results in "include everything" - so specifying an "include everything" explicitly wouldn't be necessary. But if we wanted to support this, then I think it shouldn't be an empty string, but rather eg `"/"` as that is more explicit (so if someone would specify an empty string, we could rather log a warn and skip that entry).

wdyt?

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/Configuration.java [hunk 9]
```diff
@@ -260,6 +260,27 @@
...
+                    "Any path which is added to both include & exclude paths, " +
+                    "would be removed from included paths." +
```
Comment (stefan-egli) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/Configuration.java]: Not sure we need to point out someone configuring duplicate paths specifically.

What about making this comment more generic into something like "Include and exclude paths can overlap. Exclude paths will take precedence" ?

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/Configuration.java [hunk 10]
```diff
@@ -260,6 +260,27 @@
...
+                    "Note that this value can be overridden with a system property " +
+                    "'oak.documentstore.fullGCIncludes' where paths " +
+                    "are separated with '::'. Example: -Doak.documentstore.fullGCIncludes=/content::/var")
```
Comment (stefan-egli) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/Configuration.java]: I see we sometimes state it can be overwritten via `-Doak.mongo` and sometimes `-Doak.documentstore` ... does it have this magic indeed or is one or the other wrong?

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/Configuration.java [hunk 11]
```diff
@@ -260,6 +260,27 @@
...
+
+    @AttributeDefinition(
+            name = "Full GC Exclude Paths",
+            description = "Paths which should be excluded from full Garbage collection." +
+                    "Empty value means no paths are excluded." +
+                    "Any path added to excluded list would be removed from include paths (if present)." +
+                    "Note that this value can be overridden with a system property " +
+                    "'oak.documentstore.fullGCExcludes' where paths " +
+                    "are separated with '::'. Example: -Doak.documentstore.fullGCExcludes=/content::/var")
+    String[] fullGCExcludes() default {};
```
Comment (stefan-egli) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/Configuration.java]: was wondering if the config should be called `fullGCExcludePaths` - a bit more explicit, wdyt?

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreBuilder.java [hunk 12]
```diff
@@ -170,6 +174,8 @@ public class DocumentNodeStoreBuilder<T extends DocumentNodeStoreBuilder<T>> {
     private boolean clusterInvisible;
     private boolean throttlingEnabled;
     private boolean fullGCEnabled;
+    private Set<String> fullGCIncludePaths = of();
+    private Set<String> fullGCExcludePaths = of();
```
Comment (stefan-egli) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreBuilder.java]: btw, just a side-comment, I find this less readable than say `Set.of()` (but I think this might be controversial as IDEs visualize it magically - but if you read the raw file it is less readable for me)

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreService.java (hunk 3 above)
Comment (stefan-egli) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreService.java]: Configuring the identical path in both include and exclude seems weird - but the code does behaves correctly, namely exclude has precedence, so it's "as if that path wasn't there". But I don't think we need to add an extra warn for this ..

**Review by stefan-egli — COMMENTED** [review]
lgtm in general, mainly added some comments about config names used, and config descriptions

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/Configuration.java (hunk 10 above)
Comment (rishabhdaim) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/Configuration.java]: This magic happens here: https://github.com/apache/jackrabbit-oak/blob/trunk/oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreServiceConfiguration.java#L123

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreBuilder.java (hunk 12 above)
Comment (rishabhdaim) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreBuilder.java]: fixed in https://github.com/apache/jackrabbit-oak/pull/1551/commits/55f310e2d65cfe3208ff7688ce93787eccb0cfba

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/Configuration.java (hunk 11 above)
Comment (rishabhdaim) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/Configuration.java]: fixed in https://github.com/apache/jackrabbit-oak/pull/1551/commits/55f310e2d65cfe3208ff7688ce93787eccb0cfba

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/Configuration.java (hunk 10 above)
Comment (rishabhdaim) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/Configuration.java]: fixed in https://github.com/apache/jackrabbit-oak/pull/1551/commits/55f310e2d65cfe3208ff7688ce93787eccb0cfba

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/Configuration.java (hunk 9 above)
Comment (rishabhdaim) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/Configuration.java]: fixed in https://github.com/apache/jackrabbit-oak/pull/1551/commits/55f310e2d65cfe3208ff7688ce93787eccb0cfba

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/Configuration.java (hunk 8 above)
Comment (rishabhdaim) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/Configuration.java]: fixed in https://github.com/apache/jackrabbit-oak/pull/1551/commits/55f310e2d65cfe3208ff7688ce93787eccb0cfba

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/Configuration.java (hunk 7 above)
Comment (rishabhdaim) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/Configuration.java]: fixed in https://github.com/apache/jackrabbit-oak/pull/1551/commits/55f310e2d65cfe3208ff7688ce93787eccb0cfba

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreServiceConfigurationTest.java [hunk 13]
```diff
@@ -83,7 +84,9 @@ public void defaultValues() throws Exception {
         assertEquals("MONGO", config.documentStoreType());
         assertEquals(DocumentNodeStoreService.DEFAULT_BUNDLING_DISABLED, config.bundlingDisabled());
         assertEquals(DocumentMK.Builder.DEFAULT_UPDATE_LIMIT, config.updateLimit());
-        assertEquals(Arrays.asList("/"), Arrays.asList(config.persistentCacheIncludes()));
+        assertEquals(of("/"), Arrays.asList(config.persistentCacheIncludes()));
```
Comment (stefan-egli) [inline oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreServiceConfigurationTest.java:87]: nitpick : looks unrelated/unnecessary?

**Review by stefan-egli — APPROVED** [review]
+1 with just a nitpick about an unrelated change :

**Review by Joscorbe — APPROVED** [review]

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreServiceConfigurationTest.java (hunk 13 above)
Comment (rishabhdaim) [reply oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreServiceConfigurationTest.java:87]: Yes, it can be avoided, but IntelliJ insisted on making this change.

**Review by shodaaan — APPROVED** [review]
